### PR TITLE
Task #126014 feat: added delete button/link to enable users to delete any uploaded file.

### PR DESCRIPTION
**Pull request by pravinTek**
Resolve previously added comments on praneet MR for the same feature (#126014).

### CODE REVIEW

File: administrator/views/item/tmpl/edit.php [hunk 1]
```diff
@@ -14,11 +14,13 @@
 JHtml::_('behavior.formvalidation');
 JHtml::_('formbehavior.chosen', 'select');
 JHtml::_('behavior.keepalive');
+JHtml::_('jquery.token');
 
 // Load admin language file
 $lang = JFactory::getLanguage();
 $lang->load('com_tjucm', JPATH_SITE);
 $doc = JFactory::getDocument();
+$doc->addScriptDeclaration('const site_root = "' . JUri::root() . '"');
```
[the remark this answers is not in the document]
Comment (thite-amol) [reply administrator/views/item/tmpl/edit.php]: Joomla > 3.7.0

Use below lines to get root path 

var systemPaths = Joomla.getOptions('system.paths') || {},
basePath    = systemPaths.root || '';

File: administrator/assets/js/tjucm_ajaxForm_save.js [hunk 2]
```diff
@@ -94,3 +97,39 @@ function itemformactions(tab_id, navDirection)
...
+		if(confirm(Joomla.JText._('COM_TJUCM_FILE_DELETE_CONFIRM')))
+		{
+			jQuery.ajax({
+				url: site_root + "index.php?option=com_tjucm&task=itemform.tjFileDelete",
+				type: 'POST',
+				data:{
+					filePath: filePath
+				},
+				cache: false,
+				async:true,
```
Comment (coolbung) [inline administrator/assets/js/tjucm_ajaxForm_save.js]: Why is this line needed ?

File: site/controllers/itemform.php [hunk 3]
```diff
@@ -742,4 +742,46 @@ protected function allowEdit($data = array(), $key = 'id')
...
+
+		$tjFieldsHelper = new TjfieldsHelper;
+
+		$filePath = base64_decode($filePath);
+
+		$returnValue = $tjFieldsHelper->tjFileDelete($filePath);
+
+		if ($isAjax)
+		{
+			echo (string) $returnValue;
```
Comment (coolbung) [inline site/controllers/itemform.php]: Use JResponseJson

File: administrator/assets/js/tjucm_ajaxForm_save.js [hunk 4]
```diff
@@ -94,3 +97,39 @@ function itemformactions(tab_id, navDirection)
...
+				success: function (result) {
+					if (result == '1') {
+						alert(Joomla.JText._('COM_TJUCM_FILE_DELETE_SUCCESS'));
+					}
+					else {
+						alert(Joomla.JText._('COM_TJUCM_FILE_DELETE_ERROR'));
+					}
+				},
+				complete: function(result) {
```
Comment (coolbung) [inline administrator/assets/js/tjucm_ajaxForm_save.js]: Is the different behaviour for complete and success necessary ?

File: administrator/assets/js/tjucm_ajaxForm_save.js (hunk 2 above)
Comment (praneettekdi) [reply administrator/assets/js/tjucm_ajaxForm_save.js]: @pravinTek Please remove `async:true`

File: administrator/assets/js/tjucm_ajaxForm_save.js (hunk 4 above)
Comment (praneettekdi) [reply administrator/assets/js/tjucm_ajaxForm_save.js]: @pravinTek Please try with only `success`

File: site/controllers/itemform.php (hunk 3 above)
Comment (praneettekdi) [reply site/controllers/itemform.php]: @pravinTek Please check this link and make the changes https://docs.joomla.org/JSON_Responses_with_JResponseJson

File: administrator/views/item/tmpl/edit.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (pravinTek) [reply administrator/views/item/tmpl/edit.php]: I have seen code execution and above root url is not necessary for ajax execution so, I have removed the above declaration code.
Joomla.getOptions('system.paths')  gives only project name not your domain name(Whole URL) similarly systemPaths.root also giving only root folder name.

File: administrator/assets/js/tjucm_ajaxForm_save.js (hunk 2 above)
Comment (pravinTek) [reply administrator/assets/js/tjucm_ajaxForm_save.js]: removed unwanted code for this function.

File: site/controllers/itemform.php (hunk 3 above)
Comment (pravinTek) [reply site/controllers/itemform.php]: change response logic and use JResponseJson

File: administrator/assets/js/tjucm_ajaxForm_save.js (hunk 4 above)
Comment (pravinTek) [reply administrator/assets/js/tjucm_ajaxForm_save.js]: For this functionality it's not necessary to use both .complete and .success method so, I have used only success() method.

success() and complete() methods having below difference:
    - success() only gets called if your webserver responds with a 200 OK
    - complete() will always get called
    - complete() will get called after success()

File: administrator/assets/js/tjucm_ajaxForm_save.js [hunk 5]
```diff
@@ -94,3 +93,31 @@ function itemformactions(tab_id, navDirection)
 		jQuery('#' + getTabId + ' > .active').prev('li').find('a').trigger('click');
 	}
 }
+
+/* This function deletes tjucm file via ajax */
+function deleteTjFile(filePath, fieldId)
```
[the remark this answers is not in the document]
Comment (coolbung) [reply administrator/assets/js/tjucm_ajaxForm_save.js]: cc @ankush-maherwal @manojLondhe

File: administrator/assets/js/tjucm_ajaxForm_save.js [hunk 6]
```diff
@@ -94,3 +93,31 @@ function itemformactions(tab_id, navDirection)
...
+		{
+			jQuery.ajax({
+				url: "index.php?option=com_tjucm&task=itemform.tjFileDelete",
+				type: 'POST',
+				data:{
+					filePath: filePath
+				},
+				cache: false,
+				dataType: "json",
+				success: function (result) {
```
Comment (coolbung) [inline administrator/assets/js/tjucm_ajaxForm_save.js]: Also handle delete - show a message like "Sorry the file could not be deleted, please try again"

Comment (pravinTek) [reply administrator/assets/js/tjucm_ajaxForm_save.js]: Yes sir, this scenario is already handled and showing below message.
"Error deleting file. please try again after some time."

I will change the message as you suggested.

File: site/controllers/itemform.php [hunk 7]
```diff
@@ -742,4 +742,52 @@ protected function allowEdit($data = array(), $key = 'id')
...
+		// Check for request forgeries.
+		JSession::checkToken() or jexit(JText::_('JINVALID_TOKEN'));
+
+		$app = JFactory::getApplication();
+		$jinput = $app->input;
+
+		// Here, fpht means file encoded path
+		$filePath = $jinput->get('filePath', '', 'STRING');
+
+		$isAjaxRequest = $app->input->server->getString('HTTP_X_REQUESTED_WITH', '');
```
Comment (coolbung) [inline site/controllers/itemform.php]: You already have the `$jinput` variable.

File: site/controllers/itemform.php [hunk 8]
```diff
@@ -742,4 +742,52 @@ protected function allowEdit($data = array(), $key = 'id')
...
+	public function tjFileDelete()
+	{
+		// Check for request forgeries.
+		JSession::checkToken() or jexit(JText::_('JINVALID_TOKEN'));
+
+		$app = JFactory::getApplication();
+		$jinput = $app->input;
+
+		// Here, fpht means file encoded path
+		$filePath = $jinput->get('filePath', '', 'STRING');
```
Comment (coolbung) [inline site/controllers/itemform.php]: If the value is base64 then try using the BASE64 filter instead of STRING.

Comment (pravinTek) [reply site/controllers/itemform.php]: changes made as you suggested sir.

File: site/controllers/itemform.php (hunk 7 above)
Comment (pravinTek) [reply site/controllers/itemform.php]: used already declared variable.

File: administrator/assets/js/tjucm_ajaxForm_save.js (hunk 6 above)
Comment (pravinTek) [reply administrator/assets/js/tjucm_ajaxForm_save.js]: Message changed as you suggested sir.
